gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,448 @@
+          }
+          return offsetResponse.offsets(partition.getTopicName(), partition.getId())[0];
+        } catch (Exception e) {
+          LOG.warn(String.format("Fetching offset for topic %s, partition %d has failed %d time(s). Reason: %s",
+              partition.getTopicName(), partition.getId(), i + 1, e.getMessage()));
+          if (i < NUM_TRIES_FETCH_OFFSET - 1) {
+            try {
+              Thread.sleep((long) (Math.random() * (i + 1) * 1000));
+            } catch (InterruptedException e2) {
+              LOG.error("Caught interrupted exception between retries of getting latest offsets. " + e2.getMessage());
+            }
+          }
+        }
+      }
+      throw new KafkaOffsetRetrievalFailureException(String.format(
+          "Fetching offset for topic %s, partition %d has failed.", partition.getTopicName(), partition.getId()));
+    }
+
+    @Override
+    protected Iterator<MessageAndOffset> fetchNextMessageBuffer(KafkaPartition partition, long nextOffset,
+        long maxOffset) {
+      if (nextOffset > maxOffset) {
+        return null;
+      }
+      FetchRequest fetchRequest = createFetchRequest(partition, nextOffset);
+      SimpleConsumer consumer = getSimpleConsumer(partition.getLeader().getHost(), partition.getLeader().getPort());
+
+      FetchResponse fetchResponse = null;
+      try {
+        fetchResponse = consumer.fetch(fetchRequest);
+        if (fetchResponse.hasError()) {
+          throw new RuntimeException(String.format("error code %d",
+              fetchResponse.errorCode(partition.getTopicName(), partition.getId())));
+        }
+        return getIteratorFromFetchResponse(fetchResponse, partition);
+
+      } catch (Exception e) {
+        LOG.warn(String.format(
+            "Fetch message buffer for topic %s, partition %d has failed: %s. Will refresh topic metadata and retry",
+            partition.getTopicName(), partition.getId(), e.getMessage()));
+        return refreshTopicMetadataAndRetryFetch(partition, fetchRequest);
+      }
+    }
+
+    private Iterator<MessageAndOffset> getIteratorFromFetchResponse(FetchResponse fetchResponse,
+        KafkaPartition partition) {
+      try {
+        ByteBufferMessageSet messageBuffer = fetchResponse.messageSet(partition.getTopicName(), partition.getId());
+        return messageBuffer.iterator();
+      } catch (Exception e) {
+        LOG.warn(String.format("Failed to retrieve next message buffer for topic %s, partition %d: %s."
+            + "The remainder of this partition will be skipped.", partition.getTopicName(), partition.getId(),
+            e.getMessage()));
+        return null;
+      }
+    }
+
+    private Iterator<MessageAndOffset> refreshTopicMetadataAndRetryFetch(KafkaPartition partition,
+        FetchRequest fetchRequest) {
+      try {
+        refreshTopicMetadata(partition);
+        SimpleConsumer consumer = getSimpleConsumer(partition.getLeader().getHost(), partition.getLeader().getPort());
```

Comment:
The code between here and line 369 is repeated in fetchNextMessageBuffer, can it be refactored to a different method?